models/llama3/reference_impl/generation.py

```diff
@@ -134,25 +141,29 @@ def build(
         )
         tokenizer = Tokenizer(model_path=tokenizer_path)
         assert model_args.vocab_size == tokenizer.n_words
-        if torch.cuda.is_bf16_supported():
+        if _is_cuda(device) and torch.cuda.is_bf16_supported():
             torch.set_default_tensor_type(torch.cuda.BFloat16Tensor)
-        else:
+        elif _is_cuda(device):
             torch.set_default_tensor_type(torch.cuda.HalfTensor)
+        else:
+            torch.set_default_tensor_type(torch.float16)
```

Comment:
`torch.set_default_tensor_type` seems to be deprecated starting from pytorch 2.1, see comment at API description https://pytorch.org/docs/2.5/generated/torch.set_default_tensor_type.html#torch-set-default-tensor-type

Maybe it's better to use `torch.set_default_dtype(torch.float16)`?

Reply:
Good catch! 

There are two other calls to the deprecated function: `torch.set_default_tensor_type` and no uses of `torch.set_default_dtype`. I'm not familiar with this codebase enough to know, but there may be some backwards-compatibility reasons for using this older interface (of course it may simply be a mistake!). I lean towards leaving this as is.

Reply:
I have PyTorch built with XPU backend, CUDA support not included. On such build of PyTorch `set_default_tensor_type()` does not work:
```
$ python -c "import torch; torch.set_default_tensor_type(torch.float16)"
/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/__init__.py:1204: UserWarning: torch.set_default_tensor_type() is deprecated as of PyTorch 2.1, please use torch.set_default_dtype() and torch.set_default_device() as alternatives. (Triggered internally at /pytorch/torch/csrc/tensor/python_tensor.cpp:434.)
  _C._set_default_tensor_type(t)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/dvrogozh/miniforge3/envs/llamastack-meta-reference-gpu/lib/python3.10/site-packages/torch/__init__.py", line 1204, in set_default_tensor_type
    _C._set_default_tensor_type(t)
TypeError: invalid type object: only floating-point types are supported as the default type
```

While `torch.set_default_dtype` does work fine. You can add a call to `torch.set_default_dtype` in a new non-CUDA path and leave calls to `set_default_tensor_type()` for CUDA for backward compatibility.

Reply:
```suggestion
            torch.set_default_dtype(torch.float16)
```